compose.py

```diff
@@ -0,0 +1,297 @@
+ENV TRITON_SERVER_VERSION ${{TRITON_VERSION}}
+ENV NVIDIA_TRITON_SERVER_VERSION ${{TRITON_CONTAINER_VERSION}}
+ENV TRITON_SERVER_VERSION ${{TRITON_VERSION}}
+ENV NVIDIA_TRITON_SERVER_VERSION ${{TRITON_CONTAINER_VERSION}}
+LABEL com.nvidia.tritonserver.version="${{TRITON_SERVER_VERSION}}"
+ENV PATH /opt/tritonserver/bin:${{PATH}}
+'''.format(FLAGS.upstream_container_version, FLAGS.upstream_container_version,
+           FLAGS.version, FLAGS.container_version)
+
+    # Copy over files
+    df += '''
+# Create a user that can be used to run triton as
+# non-root. Make sure that this user to given ID 1000. All server
+# artifacts copied below are assign to this user.
+ENV TRITON_SERVER_USER=triton-server
+RUN userdel tensorrt-server > /dev/null 2>&1 || true && \
+    if ! id -u $TRITON_SERVER_USER > /dev/null 2>&1 ; then \
+        useradd $TRITON_SERVER_USER; \
+    fi && \
+    [ `id -u $TRITON_SERVER_USER` -eq 1000 ] && \
+    [ `id -g $TRITON_SERVER_USER` -eq 1000 ]
+
+WORKDIR /opt/tritonserver
+RUN rm -fr /opt/tritonserver/*
+COPY --chown=1000:1000 --from=full /opt/tritonserver/LICENSE .
+COPY --chown=1000:1000 --from=full /opt/tritonserver/TRITON_VERSION .
+COPY --chown=1000:1000 --from=full /opt/tritonserver/NVIDIA_Deep_Learning_Container_License.pdf .
+COPY --chown=1000:1000 --from=full /opt/tritonserver/bin bin/
+COPY --chown=1000:1000 --from=full /opt/tritonserver/lib lib/
+COPY --chown=1000:1000 --from=full /opt/tritonserver/include/triton/core include/triton/core
+# Top-level include/core not copied so --chown does not set it correctly,
+# so explicit set on all of include
+RUN chown -R triton-server:triton-server include
+'''
+    with open(os.path.join(ddir, dockerfile_name), "w") as dfile:
+        dfile.write(df)
+
+
+### add additional backends needed
+def add_requested_backends(ddir, dockerfile_name, backends):
+    df = "# Copying over backends \n"
+    for backend in backends:
+        if (backend.lower()
+                in (EXAMPLE_BACKENDS + CORE_BACKENDS + NONCORE_BACKENDS)):
+            df += '''COPY --chown=1000:1000 --from=full /opt/tritonserver/backends/{} /opt/tritonserver/backends/{}    
+'''.format(backend, backend)
+        else:
+            log("Cannot create container from unsupported backend: " + backend)
+
+    with open(os.path.join(ddir, dockerfile_name), "a") as dfile:
+        dfile.write(df)
+
+
+def add_requested_repoagents(ddir, dockerfile_name, repoagents):
+    df = "#  Copying over repoagents \n"
+    for ra in repoagents:
+        if (ra in EXAMPLE_REPOAGENTS):
+            df += '''COPY --chown=1000:1000 --from=full /opt/tritonserver/repoagents/{} /opt/tritonserver/repoagents/{}    
+'''.format(ra, ra)
+        else:
+            log("Cannot create container from unsupported repoagent: " + ra)
+
+    with open(os.path.join(ddir, dockerfile_name), "a") as dfile:
+        dfile.write(df)
+
+
+
+# Install dependencies and run entrypoint script
+def end_gpu_dockerfile(ddir, dockerfile_name):
```

Comment:
This needs to have all the steps from build.py:create_dockerfile_linux for the production stage (excluding the copying already done above). We don't want to replicate that code so you will need to refactor build.py to group together the non-copy parts of that function to the end and then extract them to a different function. Then compose.py can import build.py and call that function.

Reply:
Thanks David. Do we need build_sha and build_id?